src/System.Management.Automation/utils/Encoding.cs

```diff
@@ -0,0 +1,387 @@
+            {
+                // If the encoding has been set to WindowsLegacy, we need to look up the actual encoding
+                if ( encoding == FileEncoding.WindowsLegacy )
+                {
+                    resolvedEncoding = GetWindowsLegacyEncoding(name);
+                }
+                else
+                {
+                    resolvedEncoding = GetEncoding(encoding);
+                }
+            }
+            else
+            {
+                // the parameter is not specifically set, so check the preference variable
+                encodingPreference = GetEncodingPreference(cmdlet.Context.SessionState);
+                // If set to unknown, we accept that it is unset
+                preferenceSetAndValid = encodingPreference != FileEncoding.Unknown; 
+                // If the encoding preference has been set to WindowsLegacy, we need to look up the actual encoding
+                if ( encodingPreference == FileEncoding.WindowsLegacy )
+                {
+                    resolvedEncoding = GetWindowsLegacyEncoding(name);
+                }
+                else if ( encodingPreference != FileEncoding.Unknown )
+                {
+                    resolvedEncoding = GetEncoding(encodingPreference);
+                }
+                // the final else would be set the encoding to GetDefaultEncoding() which was handled above
+            }
+
+            return resolvedEncoding;
+        }
+
+        // [System.Text.Encoding]::GetEncodings() | ? { $_.GetEncoding().GetPreamble() } |
+        //     Add-Member ScriptProperty Preamble { $this.GetEncoding().GetPreamble() -join "-" } -PassThru |
+        //     Format-Table -Auto
+        internal static Dictionary<String, FileEncoding> encodingMap =
+            new Dictionary<string, FileEncoding>()
+            {
+                { "255-254", FileEncoding.Unicode },
+                { "254-255", FileEncoding.BigEndianUnicode },
+                { "255-254-0-0", FileEncoding.UTF32 },
+                { "0-0-254-255", FileEncoding.BigEndianUTF32 },
+                { "239-187-191", FileEncoding.UTF8 },
+            };
+
+        internal static char[] nonPrintableCharacters = {
+            (char) 0, (char) 1, (char) 2, (char) 3, (char) 4, (char) 5, (char) 6, (char) 7, (char) 8,
+            (char) 11, (char) 12, (char) 14, (char) 15, (char) 16, (char) 17, (char) 18, (char) 19, (char) 20,
+            (char) 21, (char) 22, (char) 23, (char) 24, (char) 25, (char) 26, (char) 28, (char) 29, (char) 30,
+            (char) 31, (char) 127, (char) 129, (char) 141, (char) 143, (char) 144, (char) 157 };
+
+        internal static readonly UTF8Encoding utf8NoBom = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
+
+        // take a look at the file contents and guess at the best encoding
+        internal static FileEncoding GetEncoding(string path)
+        {
+            if (!File.Exists(path))
+            {
+                return FileEncoding.Default;
```

Comment:
not necessarily - the file could be later created, so I need to pass back something - this code was refactored from another location, so i wanted to do as little as possible